File: DarwinRealm/src/uk/ac/bournemouth/darwin/catalina/authenticator/DarwinAuthenticator.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Foobar.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package uk.ac.bournemouth.darwin.catalina.authenticator;

import net.devrieze.util.db.DBConnection;
import net.devrieze.util.db.DBConnection.DBQuery;
import net.devrieze.util.db.StringAdapter;
import org.apache.catalina.*;
import org.apache.catalina.connector.Request;
import org.apache.catalina.connector.Response;
import org.apache.catalina.deploy.LoginConfig;
import org.apache.catalina.deploy.SecurityConstraint;
import org.apache.catalina.util.LifecycleSupport;
import org.apache.catalina.valves.ValveBase;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import uk.ac.bournemouth.darwin.catalina.realm.DarwinPrincipal;
import uk.ac.bournemouth.darwin.catalina.realm.DarwinUserPrincipal;
import uk.ac.bournemouth.darwin.catalina.realm.DarwinUserPrincipalImpl;

import javax.naming.NamingException;
import javax.servlet.ServletException;
import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletResponse;
import javax.sql.DataSource;

import java.io.IOException;
import java.net.URLEncoder;
import java.security.Principal;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Iterator;
import java.util.logging.Level;
import java.util.logging.Logger;


public class DarwinAuthenticator extends ValveBase implements Authenticator, Lifecycle {


  @NotNull
  private static final String QUERY_USER_FROM_DWNID = "SELECT user FROM tokens WHERE ip=? AND token=? AND (epoch + 1800) > UNIX_TIMESTAMP()";

  private enum AuthResult {
    AUTHENTICATED,
    ERROR,
    EXPIRED,
    LOGIN_NEEDED;
  }

  private static final String AUTHTYPE = "DARWIN";

  public static final String DBRESOURCE = "java:comp/env/jdbc/webauth";

  private static final String LOGGERNAME = "DarwinRealm";

  private boolean mStarted = false;

  /**
   * The lifecycle event support for this component.
   */
  @NotNull
  protected LifecycleSupport mLifecycle = new LifecycleSupport(this);

  @Nullable
  private Context mContext;

  private String mLoginPage = "/accounts/login";


  @Override
  public void addLifecycleListener(final LifecycleListener listener) {
    mLifecycle.addLifecycleListener(listener);
  }

  @Override
  public void setContainer(final Container container) {
    super.setContainer(container);
    if (container instanceof Context) {
      mContext = (Context) container;
    }
  }

  @Override
  public LifecycleListener[] findLifecycleListeners() {
    return mLifecycle.findLifecycleListeners();
  }

  @Override
  public void removeLifecycleListener(final LifecycleListener listener) {
    mLifecycle.removeLifecycleListener(listener);
  }

  @Override
  public void start() throws LifecycleException {
    if (mStarted) {
      throw new LifecycleException("Already started");
    }
    mLifecycle.fireLifecycleEvent(START_EVENT, null);
    mStarted = true;
    setLoginPage(null); // Default is not specified.
  }

  @Override
  public void stop() throws LifecycleException {
    mLifecycle.fireLifecycleEvent(STOP_EVENT, null);
    mStarted = false;
  }

  @Override
  public void invoke(final Request request, final Response response) throws IOException, ServletException {

    final AuthResult authresult = authenticate(request);

    final Context context = this.mContext;
    final Realm realm = context == null ? null : context.getRealm();
    if (realm != null) {
      logFine("This context has an authentication realm, enforce the constraints");
      final SecurityConstraint[] constraints = realm.findSecurityConstraints(request, context);
      if (constraints == null) {
        logFine("Realm has no constraints, calling next in chain");
        // Unconstrained
        getNext().invoke(request, response);
        return;
      }
      // Need security, set cache control
      response.setHeader("Cache-Control", "private");
      if (!realm.hasUserDataPermission(request, response, constraints)) {
        denyPermission(response);
        return;
      }

      boolean authRequired = true;
      for (int i = 0; (i < constraints.length) && authRequired; i++) {
        if (!constraints[i].getAuthConstraint()) {
          authRequired = false;
        } else if (!constraints[i].getAllRoles()) {
          final String[] roles = constraints[i].findAuthRoles();
          if ((roles == null) || (roles.length == 0)) {
            authRequired = false;
          }
        }
      }

      if (authRequired) {
        if ((authresult == AuthResult.AUTHENTICATED) && realm.hasResourcePermission(request, response, constraints, context)) {
          getNext().invoke(request, response);
          return;
        } else if (authresult == AuthResult.AUTHENTICATED) { // We are authenticated, but the wrong user.
          response.sendError(HttpServletResponse.SC_FORBIDDEN, "User " + request.getUserPrincipal() + " does not have permission for "
              + realm.getInfo() + " class:" + realm.getClass().getName());
        } else {
          if (authresult != AuthResult.ERROR) {
            // Not logged in yet. So go to login page.
            final LoginConfig loginConfig = context == null ? null : context.getLoginConfig();
            String loginpage = loginConfig != null ? loginConfig.getLoginPage() : null;
            if (loginpage == null) {
              loginpage = mLoginPage;
            }
            if (loginpage != null) {
              final StringBuilder incommingPath = new StringBuilder();
              incommingPath.append(request.getPathInfo());
              response.sendRedirect(loginpage + "?redirect=" + URLEncoder.encode(incommingPath.toString(), "utf-8"));
            } else {
              response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "You need to log in for this page, but no login page is configured");
            }
          }
          return;
        }


      } else {
        getNext().invoke(request, response);
        return;
      }

    } else {
      // No realm, no authentication required.
      getNext().invoke(request, response);
    }
  }

  public String getLoginPage() {
    return mLoginPage;
  }

  public void setLoginPage(final String loginPage) {
    mLoginPage = loginPage;
  }

  public static DarwinPrincipal getPrincipal(final String user) {
    try {
      return getDarwinPrincipal(DBConnection.getDataSource(DBRESOURCE), null, user);
    } catch (NamingException e) {
      getLogger().log(Level.WARNING, "Failure to connect to database", e);
      return null;
    }
  }

  public static DarwinPrincipal asDarwinPrincipal(final Principal user) {
    try {
      final Realm realm = null;

      return toDarwinPrincipal(DBConnection.getDataSource(DBRESOURCE), realm, user);
    } catch (NamingException e) {
      getLogger().log(Level.WARNING, "Failure to connect to database", e);
      return null;
    }
  }

  private static DarwinUserPrincipal getDarwinPrincipal(final DataSource dataSource, final Realm realm, final String userName) {
    return new DarwinUserPrincipalImpl(dataSource, realm, userName);
  }

  private static DarwinUserPrincipal toDarwinPrincipal(final DataSource dataSource, final Realm realm, final Principal principal) {
    if (principal == null) {
      return null;
    }
    if (principal instanceof DarwinUserPrincipal) {
      return (DarwinUserPrincipal) principal;
    }
    return new DarwinUserPrincipalImpl(dataSource, realm, principal.getName());
  }

  @NotNull
  private static AuthResult authenticate(final Request request) {
    DarwinUserPrincipal principal;
    final DataSource dataSource;
    try {
      dataSource = DBConnection.getDataSource(DBRESOURCE);
    } catch (NamingException e) {
      getLogger().log(Level.WARNING, "Failure to connect to database", e);
      return AuthResult.ERROR;
    }
    principal = toDarwinPrincipal(dataSource, request.getContext().getRealm(), request.getUserPrincipal());
    if (principal != null) {
      logFine("Found preexisting principal, converted to darwinprincipal: " + principal.getName());
      request.setAuthType(AUTHTYPE);
      request.setUserPrincipal(principal);
      return AuthResult.AUTHENTICATED;
    }


    String user = null;
    final Cookie[] cookies = request.getCookies();
    if (cookies != null) {
      for (final Cookie cookie : cookies) {
        if ("DWNID".equals(cookie.getName())) {
          final String requestIp = request.getRemoteAddr();
          logFiner("Found DWNID cookie with value: '" + cookie.getValue() + "' and request ip:" + requestIp);
          try (final DBConnection db = DBConnection.newInstance(dataSource)){
            try (final DBQuery query = db.makeQuery(QUERY_USER_FROM_DWNID)){
              query.addParam(1, requestIp);
              query.addParam(2, cookie.getValue());
              try(final ResultSet result = query.execQuery()) {
                if (result != null) {
                  try(StringAdapter adapter = new StringAdapter(query, result, false)) {
                    final Iterator<String> it = adapter.iterator();
                    if (it.hasNext()) {
                      user = it.next();
                    } else {
                      logFine("Expired cookie: '" + cookie.getValue() + '\'');
                      return AuthResult.EXPIRED;
                    }
                  }
                }
              } catch (SQLException e) {
                throw new RuntimeException(e);
              }
            }
          }
          break;
        }
      }
    } else {
      logFiner("No authentication cookie found");
    }
    if (user != null) {
      logFine("Authenticated user " + user);
      request.setAuthType(AUTHTYPE);
      principal = getDarwinPrincipal(dataSource, request.getContext().getRealm(), user);
      request.setUserPrincipal(principal);
      return AuthResult.AUTHENTICATED;
    }
    return AuthResult.LOGIN_NEEDED;
  }

  private static void denyPermission(final Response response) throws IOException {
    response.sendError(HttpServletResponse.SC_FORBIDDEN);
  }

  private static Logger getLogger() {
    return Logger.getLogger(LOGGERNAME);
  }

  private static void logFiner(final String string) {
    getLogger().finer(string);
  }

  private static void logFine(final String string) {
    getLogger().fine(string);
  }

  private static void logInfo(final String message) {
    getLogger().info(message);
  }

}

File: DarwinRealm/src/main/kotlin/uk/ac/bournemouth/darwin/catalina/authenticator/DarwinAuthenticator.kt
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Foobar.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package uk.ac.bournemouth.darwin.catalina.authenticator

import org.apache.catalina.Authenticator
import org.apache.catalina.Container
import org.apache.catalina.Context
import org.apache.catalina.Lifecycle
import org.apache.catalina.connector.Request
import org.apache.catalina.connector.Response
import org.apache.catalina.deploy.LoginConfig
import org.apache.catalina.valves.ValveBase
import org.apache.naming.ContextBindings
import uk.ac.bournemouth.darwin.accounts.DARWINCOOKIENAME
import uk.ac.bournemouth.darwin.accounts.MAXTOKENLIFETIME
import uk.ac.bournemouth.darwin.accounts.accountDb
import uk.ac.bournemouth.darwin.catalina.realm.DarwinPrincipal
import uk.ac.bournemouth.darwin.catalina.realm.DarwinUserPrincipal
import uk.ac.bournemouth.darwin.catalina.realm.DarwinUserPrincipalImpl
import java.io.IOException
import java.io.PrintStream
import java.net.URI
import java.net.URLEncoder
import java.security.Principal
import java.util.logging.Level
import java.util.logging.Logger
import javax.naming.NamingException
import javax.servlet.ServletException
import javax.servlet.http.Cookie
import javax.servlet.http.HttpServletResponse
import javax.sql.DataSource
import javax.naming.Context as NamingContext

//@WebFilter(description = "Authentication filter for the darwin system", asyncSupported = true)
class DarwinAuthenticator : ValveBase(), Lifecycle, Authenticator {

    private enum class AuthResult {
        AUTHENTICATED,
        ERROR,
        EXPIRED,
        LOGIN_NEEDED
    }

    init {
      asyncSupported = true
    }

    var resourceName:String = DEFAULTDBRESOURCE

    var loginPage: String? = "/accountmgr/login"

    private var container_:Container? = null

    override fun getContainer() = container_
    override fun setContainer(container:Container?) {
        container_ = container
    }

    override fun initInternal() {
        super.initInternal()
        log.info("Initialising DarwinAuthenticatorValve")
    }

    fun NamingContext.print(out:PrintStream, name:String, indent:Int = 0) {
        for( binding in listBindings(name)) {
            out.print(" ".repeat(indent))
            out.println(binding.name)
            val value = binding.`object`
            if (value is NamingContext) {
                value.print(out, binding.name, indent+4)
            }
        }
    }

    val dataSource by lazy {

        val loader = ContextBindings.getClassLoader()
        try {
            val context = loader.lookup(if (container_ is Context) "comp/env/" else "") as NamingContext

            context.lookup(resourceName) as DataSource
        } catch (e:NamingException) {
            System.err.println("Failure to look up name $resourceName in context ${loader}")
            System.err.println("  the context is:")
            loader.print(System.err, "comp", 6)
            throw NamingException("Failure to look up $resourceName").initCause(e)
        }
    }



    private fun invokeNext(request: Request, response: Response) {
        next.invoke(request, response)
        val cookieNote = request.getNote(DARWINCOOKIENAME)
        if (cookieNote != null) { log.finer("Found a cookie note on the request") }
        (cookieNote as? String)?.let { token -> response.addCookie(createAuthCookie(token, request.isSecure))}
    }

    @Throws(IOException::class, ServletException::class)
    override fun invoke(request: Request, response: Response) {
        log.info("Invoking DarwinAuthenticator for ${request.method} request ${request.requestURI}")
        val container = container_!!

        // First do any handling of already present authentication information
        val authresult = lazy { authenticateHelper(dataSource, request, response) }
        request.setNote("response", response)
        run {
            if ((container is Context && container.preemptiveAuthentication) || (request.cookies?.any { c -> c.name==DARWINCOOKIENAME } ?: false)) { authresult.value }
            // If the context wants us to do preemptive authentication, make sure to get the authentication result.
            // That has side-effects that will register the principal.
        }

        loginPage?.let { // If the target is the login page, now just go there.
            val loginPageUrl = URI.create(it)
            if (request.requestURI==loginPageUrl.path) {
                invokeNext(request, response)
                return
            }
        }

        val realm = container.realm
        if (realm != null) {
            log.info("DarwinAuthenticator: this context has an authentication realm, enforce the constraints")
            val constraints = (container as? Context)?.let { realm.findSecurityConstraints(request, it) }
            if (constraints == null) {
                log.fine ("Realm has no constraints, calling next in chain")
                // Unconstrained
                invokeNext(request, response)
                return
            }
            // Need security, set cache control
            response.setHeader("Cache-Control", "private")
            if (!realm.hasUserDataPermission(request, response, constraints)) {
                when (authresult.value) {
                    AuthResult.AUTHENTICATED -> response.sendError(HttpServletResponse.SC_FORBIDDEN, "Forbidden")
                    AuthResult.ERROR -> response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Authentication error")
                    else -> requestLogin(request, response)
                }
                return
            }

            var authRequired: Boolean
            if (constraints == null) {
                authRequired = false
            } else {
                authRequired = true
                for(constraint in constraints) {
                    if (!constraint.authConstraint) {
                        authRequired = false
                        break;
                    } else if (!constraint.allRoles) {
                        val roles = constraint.findAuthRoles()
                        if (roles == null || roles.size == 0) {
                            authRequired = false
                            break;
                        }
                    }
                }
            }

            if (authRequired) {
                when (authresult.value) {
                    AuthResult.AUTHENTICATED -> if (container is Context && realm.hasResourcePermission(request, response, constraints, container)) {
                            invokeNext(request, response)
                        } else {
                            response.sendError(HttpServletResponse.SC_FORBIDDEN, "User " + request.userPrincipal + " does not have permission for " + realm.info + " class:" + realm.javaClass.name)
                        }
                    AuthResult.LOGIN_NEEDED, AuthResult.EXPIRED -> requestLogin(request, response)
                    else -> {
                        response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Authentication required, but encountered an error")
                    }
                }
            } else {
                invokeNext(request, response)
            }

        } else {
            log.info("No realm set for the request")
            // No realm, no authentication required.
            invokeNext(request, response)
        }
    }

    private fun requestLogin(request: Request, response: HttpServletResponse) {
        val loginPage = this.loginPage
        val decodedRequestURI: String? = request.decodedRequestURI
        if (loginPage != null && decodedRequestURI!=null) {
            response.sendRedirect("${loginPage}?redirect=${URLEncoder.encode(decodedRequestURI, "utf-8")}")
        } else {
            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "You need to log in for this page, but no login page is configured")
        }
    }

    val authMethod: String get() = AUTHTYPE

    /**
     * Overridden version of login that does not use register. We don't need tomcat session cookies, and we certainly
     * don't carry passwords around.
     */
    override fun login(userName: String, password: String, request: Request) {
        log.info("Login requested for ${request.requestURI}. Passing to realm")
        val principal = container_!!.realm.authenticate(userName, password) ?: throw ServletException("Invalid credentials")

        accountDb(dataSource) {
            val authtoken = this.createAuthtoken(userName, request.remoteAddr)

            val response = request.getNote("response") as? Response
            if (response!=null) {
                log.log(Level.FINE, "Setting a cookie on the response")
                response.addCookie(createAuthCookie(authtoken, request.isSecure))
            } else {
                log.info("Setting a note for the cookie as there is no recorded response")
                request.setNote(DARWINCOOKIENAME, authtoken)
            }
        }

        if (log.isLoggable(Level.FINER)) {
            val name = if (principal == null) "none" else principal.name
            log.finer("Authenticated '$name' with type '${AUTHTYPE}'")
        }

        // Set the user into the request
        request.userPrincipal = principal
    }

    @Throws(ServletException::class)
    override fun logout(request: Request) {
        log.fine("Logging out")
        (request.getNote("response") as? HttpServletResponse) ?.let { response ->
            clearCookies(request, response)
        }

        request.userPrincipal = null
    }

    override fun authenticate(request: Request, response: HttpServletResponse): Boolean {
        val authResult = authenticateHelper(dataSource, request, response)
        when (authResult) {
            AuthResult.AUTHENTICATED -> return true
            else -> {
                requestLogin(request, response)
                return false
            } // We have sent the correct message
        }
    }

    @Throws(IOException::class)
    override fun authenticate(request: Request, response: HttpServletResponse, config: LoginConfig): Boolean {
        return authenticate(request, response)
    }

    companion object {

        private val AUTHTYPE = "DARWIN"

        val DEFAULTDBRESOURCE = "webauthadm"

        private val log = Logger.getLogger(DarwinAuthenticator::class.java.name)

        fun Principal.asDarwinPrincipal(dataSource: DataSource): DarwinPrincipal {
            try {
                return toDarwinPrincipal(dataSource, this)
            } catch (e: NamingException) {
                log.log(Level.WARNING, "Failure to connect to database", e)
                throw e
            }

        }

        private fun toDarwinPrincipal(dataSource: DataSource, principal: Principal): DarwinUserPrincipal {
            if (principal is DarwinUserPrincipal) {
                return principal
            }
            return DarwinUserPrincipalImpl(dataSource, principal.name)
        }

        private fun createAuthCookie(authtoken: String, secureCookie:Boolean) = Cookie(DARWINCOOKIENAME, authtoken).apply {
            maxAge = MAXTOKENLIFETIME;
            path="/"
            secure = secureCookie
            version=1
        }

        private fun clearCookies(request: Request, response:HttpServletResponse) {
            request.removeNote(DARWINCOOKIENAME)
            request.cookies?.asSequence()?.filter ({ it.name == DARWINCOOKIENAME })?.forEach { staleCookie ->
                staleCookie.maxAge = 0
                staleCookie.value = ""
                staleCookie.path="/"
                staleCookie.secure = request.isSecure
                response.addCookie(staleCookie)
            }

        }

        private fun authenticateHelper(dataSource: DataSource, request: Request, response: HttpServletResponse): AuthResult {
            val origPrincipal: Principal? = request.userPrincipal

            val authTokens: List<String> =
                  ((request.getHeader(DARWINCOOKIENAME)?.let{ listOf(it) }?: emptyList()).asSequence() +
                  (request.cookies?.asSequence()?.filter ({ it.name == DARWINCOOKIENAME })?.map { it.value }?: emptySequence())).toList()

            if (authTokens.isEmpty())
                    return AuthResult.LOGIN_NEEDED.apply { log.info("authenticateHelper: No user found") }

            if (origPrincipal != null) {
                if (origPrincipal !is DarwinUserPrincipal) {
                    log.info("Found preexisting principal, converted to darwinprincipal: ${origPrincipal.name}")
                    request.authType = AUTHTYPE
                    request.userPrincipal = origPrincipal.asDarwinPrincipal(dataSource)
                }

                return AuthResult.AUTHENTICATED.apply { log.info("authenticateHelper: previously authenticated as ${request.userPrincipal.name}") }
            }

            try {

                accountDb(dataSource) {
                    cleanAuthTokens() // First get rid of no longer needed tokens

                    // Try all cookies, not the first
                    val userInfo = authTokens.asSequence().map { authToken ->
                        userFromToken(authToken, request.remoteAddr)?.let { user -> authToken to user }
                    }.filterNotNull().firstOrNull()

                    if (userInfo != null) {
                        val user = userInfo.second
                        val authToken = userInfo.first
                        request.userPrincipal = DarwinUserPrincipalImpl(dataSource, user, getUserRoles(user))
                        // Set the cookie as a note so it can be removed for example in logout
                        request.setNote(DARWINCOOKIENAME, authToken)

                        return AuthResult.AUTHENTICATED.apply { log.info("authenticateHelper: authenticated as ${request.userPrincipal.name}") }
                    }
                }
                // invalidate all old cookies
                clearCookies(request, response)

                return AuthResult.LOGIN_NEEDED.apply { log.info("authenticateHelper: cookie no longer valid (${authTokens.first()})") }

            } catch (e: Exception) {
                log.log(Level.WARNING, "Failure in verifying user", e)
                return AuthResult.ERROR
            }

        }

    }

}
